pallets/pallet-assets-handler/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+			amount: BalanceOf<T>,
+		},
+		TokenBridgeOut {
+			asset_id: Option<AssetId<T>>,
+			to: T::AccountId,
+			// Before Fee
+			amount: BalanceOf<T>,
+			fee: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidResourceId,
+		CannotPayAsFee,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Stores an asset id on chain under an associated resource ID.
+		#[pallet::call_index(0)]
+		#[pallet::weight({1000})]
+		pub fn set_resource(
+			origin: OriginFor<T>,
+			resource_id: ResourceId,
+			asset: AssetInfo<AssetId<T>, BalanceOf<T>>,
+		) -> DispatchResult {
+			<T as pallet_bridge::Config>::BridgeCommitteeOrigin::ensure_origin(origin)?;
+			ResourceToAssetInfo::<T>::insert(resource_id, asset.clone());
+			Self::deposit_event(Event::ResourceUpdated { resource_id, asset });
+			Ok(())
+		}
+
+		/// Removes a resource ID from the resource mapping.
+		///
+		/// After this call, bridge transfers with the associated resource ID will
+		/// be rejected.
+		#[pallet::call_index(1)]
+		#[pallet::weight({1000})]
+		pub fn remove_resource(origin: OriginFor<T>, resource_id: ResourceId) -> DispatchResult {
+			<T as pallet_bridge::Config>::BridgeCommitteeOrigin::ensure_origin(origin)?;
+			ResourceToAssetInfo::<T>::remove(resource_id);
+			Self::deposit_event(Event::ResourceRemoved { resource_id });
+			Ok(())
+		}
+	}
+}
+
+impl<T, B, A> BridgeHandler<B, A, ResourceId> for Pallet<T>
+where
+	T: Config
+		+ frame_system::Config<AccountId = A>
+		+ pallet_bridge::Config<Balance = B>
+		+ pallet_assets::Config<Balance = B>
+		+ pallet_balances::Config<Balance = B>,
+	B: Copy + FixedPointOperand + CheckedSub,
+	A: Clone,
+{
+	fn prepare_token_bridge_in(
+		resource_id: ResourceId,
+		who: A,
+		amount: B,
+	) -> Result<B, DispatchError> {
+		let asset_info = Self::resource_to_asset_info(resource_id);
+		match asset_info {
+			None => Err(Error::<T>::InvalidResourceId.into()),
+			// Native token
+			Some(AssetInfo { fee: _, asset: None }) => {
+				Self::deposit_event(Event::TokenBridgeIn {
+					asset_id: None,
+					to: who.clone(),
+					amount,
+				});
+				pallet_balances::Pallet::<T>::mint_into(&who, amount)
+			},
+			// pallet assets
+			Some(AssetInfo { fee: _, asset: Some(asset) }) => {
+				Self::deposit_event(Event::TokenBridgeIn {
+					asset_id: Some(asset.clone()),
+					to: who.clone(),
+					amount,
+				});
+				pallet_assets::Pallet::<T>::mint_into(asset, &who, amount)
+			},
+		}
+	}
+	// Return actual amount to target chain after deduction e.g fee
+	fn prepare_token_bridge_out(
+		resource_id: ResourceId,
+		who: A,
+		amount: B,
+	) -> Result<B, DispatchError> {
+		let asset_info = Self::resource_to_asset_info(resource_id);
+		match asset_info {
+			None => Err(Error::<T>::InvalidResourceId.into()),
+			// Native token
+			Some(AssetInfo { fee, asset: None }) => {
+				Self::deposit_event(Event::TokenBridgeOut {
+					asset_id: None,
+					to: who.clone(),
+					amount,
+					fee,
+				});
```

Comment:
Is this fn transactional? Normally we store the event at the end after database write is done

Reply:
Yes, it is already transactional. I apply the transacitonal macro as the extrinsic which use it.

This is an atomic inner function. Do you suggest we use transactional macro apply to it directly?

Reply:
I think the extrinsic is defaulted to be transactional - but not sure about the trait impl though.

Ideally we could have some tests for it... (or play it safe and put the event deposit at the very bottom)